### ISSUE TITLE
invalid DB query

### ISSUE THREAD

**Issue by webeks** [rainlab/translate-plugin#174]
When using PostgreSql with latest translate-plugin and blog there is an exception thrown

`SQLSTATE[42883]: Undefined function: 7 ERROR: operator does not exist: integer = character varying LINE 1: ...b_translate_indexes" on "rainlab_blog_posts"."id" = "rainlab... ^ HINT: No operator matches the given name and argument type(s). You might need to add explicit type casts. (SQL: select "rainlab_blog_posts".* from "rainlab_blog_posts" left join "rainlab_translate_indexes" on "rainlab_blog_posts"."id" = "rainlab_translate_indexes"."model_id" and "rainlab_translate_indexes"."model_type" = RainLab\Blog\Models\Post and "rainlab_translate_indexes"."locale" = en where ("rainlab_blog_posts"."slug" = first-blog-post or ("rainlab_translate_indexes"."item" = slug and "rainlab_translate_indexes"."value" = first-blog-post)) and "published" is not null and "published" = 1 and "published_at" is not null and "published_at" < 2016-10-11 14:29:10 limit 1)
`

Please check **plugins/rainlab/translate/behaviors/TranslatableModel.php** L55

`->on($this->model->getQualifiedKeyName(), '=', 'rainlab_translate_indexes.model_id')`

Basically you are joining integer on varchar field which is not allowed without type casting.

**Comment by webeks** [comment]
I am looking into this. As far as I see it we can either check DB type and cast accordingly. Or enforce compatibility for all translatable models to use primary key as integer by modifying model_id to become integer instead of string. 

I don't particular like either solution. Do you have any suggestions?

**Comment by daftspunk** [comment]
We have a solution for this already...
```
`DbDongle::cast('model_id', 'TEXT');
```

... however I'm not seeing how we arrive at this problem. As per the contribution guidelines, please provide the exact steps you are taking for the issue to occur. This will be used for peer review.

**Comment by webeks** [comment]
Steps to recreate:

1. install extension on PostgreSql
2. open any blog post on frontend

**Comment by daftspunk** [comment]
Reference when fixed: #176

**Comment by daftspunk** [comment]
Unable to replicate

1. Running PostgreSQL and latest code base (Build 373+)
1. Two languages set up with Translate plugin (Russian - ru / English - en)
1. Navigate to blog area
1. Click on a blog post
1. Blog post opens successfully

**Comment by Eoler** [comment]
> Should I fetch core October directly from git as well to get to build 373?

No, changing `edgeUpdates cms.config` to true and checking for updates in backend should be enough.

**Comment by webeks** [comment]
No, still not working. 

* clean install (via command line composer / php artisan)
* default theme with added blog CMS page (see below)
* build 376
* tested on build 380 with edgeUpdates as well
* blog v. 1.2.14
* tranlsate v. 1.2.6
* using PHP 7.0.2-4+deb.sury.org~trusty+1
* postgresql 9.5.0 (PostgreSQL(libpq) Version 9.3.14)

Can I provide something else to identify exact process to reproduce the issue?
 

`SQLSTATE[42883]: Undefined function: 7 ERROR: operator does not exist: integer = character varying LINE 1: ...b_translate_indexes" on "rainlab_blog_posts"."id" = "rainlab... ^ HINT: No operator matches the given name and argument type(s). You might need to add explicit type casts. (SQL: select "rainlab_blog_posts".* from "rainlab_blog_posts" left join "rainlab_translate_indexes" on "rainlab_blog_posts"."id" = "rainlab_translate_indexes"."model_id" and "rainlab_translate_indexes"."model_type" = RainLab\Blog\Models\Post and "rainlab_translate_indexes"."locale" = en where ("rainlab_blog_posts"."slug" = first-blog-post or ("rainlab_translate_indexes"."item" = slug and "rainlab_translate_indexes"."value" = first-blog-post)) and "published" is not null and "published" = 1 and "published_at" is not null and "published_at" < 2016-11-10 14:03:00 limit 1)
`
/mnt/hgfs/WWW/october-demo/vendor/laravel/framework/src/Illuminate/Database/Connection.php line 662

**Comment by webeks** [comment]
@daftspunk we need to cast this https://github.com/rainlab/translate-plugin/blob/e7ce25baf1006443d6309a13037d0a3edb79e912/behaviors/TranslatableModel.php#L55

Any idea how to do it?

**Comment by daftspunk** [comment]
@webeks try updating this line to this and test if it works:

```
->on(\DbDongle::cast($this->model->getQualifiedKeyName(), 'TEXT'), '=', 'rainlab_translate_indexes.model_id')
```

Once confirmed we can add the code in.

**Comment by webeks** [comment]
No, this produces another error:
`SQLSTATE[42601]: Syntax error: 7 ERROR: syntax error at or near "as" LINE 1: ...nslate_indexes" on "CAST(rainlab_blog_posts"."id" as "TEXT)"... ^ (SQL: select "rainlab_blog_posts".* from "rainlab_blog_posts" left join "rainlab_translate_indexes" on "CAST(rainlab_blog_posts"."id" as "TEXT)" = "rainlab_translate_indexes"."model_id" and "rainlab_translate_indexes"."model_type" = RainLab\Blog\Models\Post and "rainlab_translate_indexes"."locale" = en where ("rainlab_blog_posts"."slug" = first-blog-post or ("rainlab_translate_indexes"."item" = slug and "rainlab_translate_indexes"."value" = first-blog-post)) and "published" is not null and "published" = 1 and "published_at" is not null and "published_at" < 2016-11-14 13:26:40 limit 1)
`

**Comment by webeks** [comment]
this seems to be working
`  ->on(\DB::raw(\DbDongle::cast($this->model->getQualifiedKeyName(), 'TEXT')), '=', 'rainlab_translate_indexes.model_id')
`

**Comment by daftspunk** [comment]
The above [working code] opens up a small security attack vector.

Can you give this one a try too...

```
->on('rainlab_translate_indexes.model_id', '=', \DbDongle::cast($this->model->getQualifiedKeyName(), 'TEXT'))
```

**Comment by webeks** [comment]
No :) 

`SQLSTATE[42601]: Syntax error: 7 ERROR: syntax error at or near "as" LINE 1: ...exes"."model_id" = "CAST(rainlab_blog_posts"."id" as "TEXT)"... ^ (SQL: select "rainlab_blog_posts".* from "rainlab_blog_posts" left join "rainlab_translate_indexes" on "rainlab_translate_indexes"."model_id" = "CAST(rainlab_blog_posts"."id" as "TEXT)" and "rainlab_translate_indexes"."model_type" = RainLab\Blog\Models\Post and "rainlab_translate_indexes"."locale" = en where ("rainlab_blog_posts"."slug" = first-blog-post or ("rainlab_translate_indexes"."item" = slug and "rainlab_translate_indexes"."value" = first-blog-post)) and "published" is not null and "published" = 1 and "published_at" is not null and "published_at" < 2016-11-15 13:25:47 limit 1)
`

Problem is that CAST is getting escaped if we run it without raw.  
`"model_id" = "CAST(rainlab_blog_posts"."id" as "TEXT)"`
